File: typescript/packages/node/src/core/ssh/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { IndexCacheStore } from '@struktoai/mirage-core/cache/index/store'
import type { PathSpec } from '@struktoai/mirage-core/types'
import { enoent } from '@struktoai/mirage-core/utils/errors'
import type { SSHAccessor } from '../../accessor/ssh.ts'
import { readChunks } from './stream.ts'
import { isNoSuchFile, joinRoot, stripPrefix } from './utils.ts'

/**
 * Read a file, optionally only a byte range of it.
 *
 * SFTP seeks, so a window costs the window rather than the whole file:
 * `start`/`end` bound the READs the stream issues. `end` is inclusive,
 * matching node's own `createReadStream`.
 *
 * Args:
 *   accessor: SSH accessor.
 *   p: the path to read.
 *   _index: unused; SFTP resolves the path itself.
 *   options: `{offset, size}`, the byte window, or absent for the whole file.
 */
export async function read(
  accessor: SSHAccessor,
  p: PathSpec,
  _index?: IndexCacheStore,
  options?: { offset?: number; size?: number },
): Promise<Uint8Array> {
  const start = options?.offset ?? 0
  const size = options?.size ?? null
  const sftp = await accessor.sftp()
  const virtual = stripPrefix(p)
  const remote = joinRoot(accessor.config.root ?? '/', virtual)
  // ssh2's readFile stats the file and issues a single READ for the whole
  // size; servers that honor large reads (asyncssh) reply with a packet over
  // ssh2's 256 KiB cap, a fatal protocol error that strands the pending
  // callback. Stream in bounded chunks instead.
  const rs = sftp.createReadStream(remote, {
    start,
    ...(size === null ? {} : { end: start + size - 1 }),
  })
  const chunks: Uint8Array[] = []
  let total = 0
  try {
    for await (const u8 of readChunks(rs)) {
      chunks.push(u8)
      total += u8.byteLength
    }
  } catch (err) {
    if (isNoSuchFile(err)) throw enoent(p)
    throw err
  }
  const out = new Uint8Array(total)
  let at = 0
  for (const c of chunks) {
    out.set(c, at)
    at += c.byteLength
  }
  return out
}

File: python/mirage/core/mem0/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator
from typing import Any

from mirage.accessor.mem0 import Mem0Accessor
from mirage.cache.index import NULL_INDEX, IndexCacheStore
from mirage.core.hierarchy.read import make_read
from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.mem0.client import get_memory
from mirage.core.mem0.scope import detect_scope
from mirage.core.render.json import json_bytes
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def _resolve_memory(
    accessor: Mem0Accessor,
    path: PathSpec,
    index: IndexCacheStore,
) -> dict[str, Any]:
    match = detect_scope(path)
    if match.kind != "memory":
        raise enoent(path)
    lookup = await index.get(path.virtual)
    cached = (lookup.entry.extra.get("memory")
              if lookup.entry is not None else None)
    if isinstance(cached, dict):
        return cached
    return await get_memory(accessor.client, match.slots["memory_id"], path)


async def _read_memory(accessor: Mem0Accessor, match: ScopeMatch,
                       path: PathSpec, index: IndexCacheStore) -> bytes:
    memory = await _resolve_memory(accessor, path, index)
    return json_bytes(memory)


read = make_read(detect_scope, {"memory": _read_memory})


async def read_stream(
    accessor: Mem0Accessor,
    path: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> AsyncIterator[bytes]:
    """Stream a memory as full JSON bytes (used by jq).

    Args:
        accessor (Mem0Accessor): mem0 accessor.
        path (PathSpec): the memory file path.
        index (IndexCacheStore): index cache.
    """
    memory = await _resolve_memory(accessor, path, index)
    yield json_bytes(memory)
